Fails typescript build with `skipLibCheck: false` as of typescript 4.4.2

## 🐛 Bug report

### Current behavior

As of `typescript@4.4.2`, projects with the `skipLibCheck` option disabled will see errors when running `tsc` with `reakit` as a dependency.

### Steps to reproduce the bug

1. Clone [minimal repo](https://github.com/dkempner/reakit-skip-lib-check)
2. `yarn`
3. `yarn run tsc`
4. Notice typescript errors from `reakit/ts/Form`

```
node_modules/reakit/ts/Form/__utils/types.d.ts:14:5 - error TS2411: Property '["0"]' of type 'keyof T | undefined' is not assignable to 'number' index type 'string | number'.

14     ["0"]?: keyof T;
       ~~~~~

node_modules/reakit/ts/Form/__utils/types.d.ts:15:5 - error TS2411: Property '["1"]' of type '(P extends { 0: infer K0; } ? K0 extends keyof T ? keyof T[K0] : never : never) | undefined' is not assignable to 'number' index type 'string | number'.

15     ["1"]?: P extends {
       ~~~~~

node_modules/reakit/ts/Form/__utils/types.d.ts:18:5 - error TS2411: Property '["2"]' of type '(P extends { 0: infer K0; 1: infer K1; } ? K0 extends keyof T ? K1 extends keyof T[K0] ? keyof T[K0][K1] : never : never : never) | undefined' is not assignable to 'number' index type 'string | number'.

18     ["2"]?: P extends {
       ~~~~~

node_modules/reakit/ts/Form/__utils/types.d.ts:22:5 - error TS2411: Property '["3"]' of type '(P extends { 0: infer K0; 1: infer K1; 2: infer K2; } ? K0 extends keyof T ? K1 extends keyof T[K0] ? K2 extends keyof T[K0][K1] ? keyof T[K0][K1][K2] : never : never : never : never) | undefined' is not assignable to 'number' index type 'string | number'.

22     ["3"]?: P extends {
       ~~~~~

node_modules/reakit/ts/Form/__utils/types.d.ts:27:5 - error TS2411: Property '["4"]' of type '(P extends { 0: infer K0; 1: infer K1; 2: infer K2; 3: infer K3; } ? K0 extends keyof T ? K1 extends keyof T[K0] ? K2 extends keyof T[K0][K1] ? K3 extends keyof T[K0][K1][K2] ? keyof T[K0][K1][K2][K3] : never : never : never : never : never) | undefined' is not assignable to 'number' index type 'string | number'.

27     ["4"]?: P extends {
       ~~~~~

node_modules/reakit/ts/Form/__utils/types.d.ts:33:5 - error TS2411: Property '["5"]' of type '(P extends { 0: infer K0; 1: infer K1; 2: infer K2; 3: infer K3; 4: infer K4; } ? K0 extends keyof T ? K1 extends keyof T[K0] ? K2 extends keyof T[K0][K1] ? K3 extends keyof T[K0][K1][K2] ? K4 extends keyof T[K0][K1][K2][K3] ? keyof T[K0][K1][K2][K3][K4] : never : never : never : never : never : never) | undefined' is not assignable to 'number' index type 'string | number'.

33     ["5"]?: P extends {
       ~~~~~
```

### Expected behavior

Pass `tsc` without having to enable `skipLibCheck`.

### Possible solutions

Fix the typescript definitions in `Form/__utils/types.d.ts` with `typescript@4.4.2`.

### Environment

Please, run the command below inside your project directory.
```sh
$ npx envinfo --system --binaries --browsers --npmPackages "{react*,reakit*}"

  System:
    OS: macOS 11.5.2
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 84.11 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.17.4 - ~/.nvm/versions/node/v14.17.4/bin/node
    Yarn: 1.22.11 - ~/.nvm/versions/node/v14.17.4/bin/yarn
    npm: 6.14.14 - ~/.nvm/versions/node/v14.17.4/bin/npm
  Browsers:
    Chrome: 92.0.4515.159
    Firefox: 84.0.1
  npmPackages:
    react: 17.0.2 => 17.0.2
    react-dom: 17.0.2 => 17.0.2
    reakit: 1.3.6 => 1.3.6
```

--- a/packages/reakit/src/Form/__utils/types.ts
+++ b/packages/reakit/src/Form/__utils/types.ts
@@ -11,7 +11,7 @@ export type ArrayWithLength<N extends number> = { [K in N]: any };
  * @template T Object with keys { foo: { bar: [{ baz }] } }
  * @template P Path ["foo", "bar", 0, "baz"]
  */
-export interface DeepPathArray<T, P> extends ReadonlyArray<string | number> {
+export interface DeepPathArray<T, P> extends ReadonlyArray<any> {
   ["0"]?: keyof T;
   ["1"]?: P extends {
     ["0"]: infer K0;

